src/main/flight/autopilot.c

```diff
@@ -91,14 +155,209 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        normalisedErrorAngle -= 360.0f; // Range: -180 to 180
+    }
+
+    // Calculate distance error proportions for pitch and roll
+    const float errorAngleRadians = normalisedErrorAngle * (M_PIf / 180.0f);
+    const float rollProportion = -sin_approx(errorAngleRadians); // + 1 when target is left, -1 when to right, of the craft
+    const float pitchProportion = cos_approx(errorAngleRadians); // + 1 when target is ahead, -1 when behind, the craft
+
+    // set the filter gain used for D and J
+    // TO DO - maybe use fixed at GPS data rate?
+    const float gain = pt1FilterGain(positionLpfCutoffHz, gpsDataIntervalS);
+
+    // ** calculate P, D and J for pitch and roll axes, independently.
+    // TODO for loop by axis
+
+    // roll
+    const float distanceRoll = rollProportion * posHold.distanceCm;
+    // positive distances mean nose towards target, should roll forward (positive roll)
+
+    // we need separate velocity for roll so the filter lag isn't problematic
+    float velocityRoll = (distanceRoll - posHold.previousDistanceRoll) * gpsDataIntervalHz;
+    posHold.previousDistanceRoll = distanceRoll;
+    // lowpass filter the velocity
+    pt1FilterUpdateCutoff(&velocityRollLpf, gain);
+    velocityRoll = pt1FilterApply(&velocityRollLpf, velocityRoll);
+
+    float accelerationRoll = (velocityRoll - posHold.previousVelocityRoll) * gpsDataIntervalHz; // positive when moving away
+    posHold.previousVelocityRoll = velocityRoll;
+    // lowapss filter the acceleration value again, effectively PT2, it's very noisy
+    pt1FilterUpdateCutoff(&accelerationRollLpf, gain);
+    accelerationRoll = pt1FilterApply(&accelerationRollLpf, accelerationRoll);
+
+    const float rollP = distanceRoll * positionPidCoeffs.Kp;
+    const float rollD = velocityRoll * positionPidCoeffs.Kd;
+    const float rollJ = accelerationRoll * positionPidCoeffs.Kf;
+
+    // pitch
+    const float distancePitch = pitchProportion * posHold.distanceCm;
+    // positive distances mean nose towards target, should pitch forward (positive pitch)
+
+    float velocityPitch = (distancePitch - posHold.previousDistancePitch) * gpsDataIntervalHz;
+    posHold.previousDistancePitch = distancePitch;
+    // lowpass filter the velocity
+    pt1FilterUpdateCutoff(&velocityPitchLpf, gain);
+    velocityPitch = pt1FilterApply(&velocityPitchLpf, velocityPitch);
+
+    float accelerationPitch = (velocityPitch - posHold.previousVelocityPitch) * gpsDataIntervalHz; // positive when moving away
+    posHold.previousVelocityPitch = velocityPitch;
+    // lowapss filter the acceleration value again, effectively PT2, it's very noisy
+    pt1FilterUpdateCutoff(&accelerationPitchLpf, gain);
+    accelerationPitch = pt1FilterApply(&accelerationPitchLpf, accelerationPitch);
+
+    const float pitchP = distancePitch * positionPidCoeffs.Kp;
+    const float pitchD = velocityPitch * positionPidCoeffs.Kd;
+    const float pitchJ = accelerationPitch * positionPidCoeffs.Kf;
```

Comment:
could this code be refactored to use same pattern as in pid.c: iterate through the axes (could be angle_index_t) and update posPID[axis].P etc. This will shorten the method. Also it will be easier to separate this code to use one PID implementation for all needs.

Reply:
Yes I was thinking that there should be a for.. loop through the axes.  I just wasn't sure how to do that (not very clever with that sort of thing) and did it the 'easy way' first.

Maybe in a later PR we can refactor this more efficiently?

Also using the pid layout means the label Kf gets applied to the Acceleration/Jerk value... it's OK though not ideal.